### PR TITLE
Restructure project and add Storybook

### PR DESCRIPTION
**Pull request by MythicManiac**

### CODE REVIEW

File: packages/components/src/components/PackageUpload.tsx [hunk 1]
```diff
@@ -139,9 +139,15 @@ const PackageUploadForm: React.FC<PackageUploadFormProps> = ({
...
-                    )}
+                    {errors.communities &&
+                      errors.communities.map(
+                        (value, index) =>
+                          value && (
+                            <FormErrorMessage key={index}>
+                              {value.message}
+                            </FormErrorMessage>
+                          )
+                      )}
```
Comment (nihaals) [inline packages/components/src/components/PackageUpload.tsx:150]: This should check if `errors.communities.*.message` is an actual value

File: packages/nextjs/next.config.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+        rule.include = [
+          ...rule.include,
+          // eslint-disable-next-line @typescript-eslint/no-var-requires
+          require("path").join(__dirname, ".."),
+        ];
+      }
+    });
+    return config;
+  },
+};
```
Comment (nihaals) [inline packages/nextjs/next.config.js:14]: Why is this needed? Some comments would help explain (I'm assuming it's related to importing from the components library)

File: packages/nextjs/package.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "@thunderstore/nextjs",
+  "version": "0.1.0",
+  "private": true,
+  "repository": "https://github.com/thunderstore-io/thunderstore-ui/tree/master/packages/nextjs",
```
Comment (nihaals) [inline packages/nextjs/package.json:5]: From the [docs](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#repository):
> The URL should be a publicly available (perhaps read-only) url that can be handed directly to a VCS program without any modification. It should not be a url to an html project page that you put in your browser. It's for computers.

A relevant example:
```json
{
  "repository": {
    "type": "git",
    "url": "https://github.com/facebook/react.git",
    "directory": "packages/react-dom"
  }
}
```

File: packages/nextjs/package.json [hunk 4]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "@thunderstore/nextjs",
+  "version": "0.1.0",
```
Comment (nihaals) [inline packages/nextjs/package.json:3]: Is this intentionally `0.1.0` or should it be unversioned (`0.0.0`)?

File: packages/nextjs/package.json [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    "build": "next build",
+    "start": "next start",
+    "typecheck": "tsc --noEmit"
+  },
+  "dependencies": {
+    "@thunderstore/components": "*",
+    "next": "^10.1.3",
+    "react": "^17.0.2",
+    "react-dom": "^17.0.2"
+  }
```
Comment (nihaals) [inline packages/nextjs/package.json:17]: Shouldn't some of the dev dependencies like `@types/react` be duplicated here?

File: package.json [hunk 6]
```diff
@@ -2,25 +2,17 @@
   "private": true,
   "repository": "https://github.com/thunderstore-io/thunderstore-ui",
```
Comment (nihaals) [inline package.json:3]: From the [docs](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#repository):
> The URL should be a publicly available (perhaps read-only) url that can be handed directly to a VCS program without any modification. It should not be a url to an html project page that you put in your browser. It's for computers.

This should probably end in `.git`

File: packages/components/package.json [hunk 7]
```diff
@@ -1,13 +1,14 @@
 {
-  "name": "thunderstore-components",
+  "name": "@thunderstore/components",
   "version": "0.1.0",
   "description": "Shared components for Thunderstore",
-  "repository": "https://github.com/thunderstore-io/thunderstore-ui/tree/master/thunderstore-components",
-  "main": "./dist/index.js",
+  "repository": "https://github.com/thunderstore-io/thunderstore-ui/tree/master/packages/components",
```
Comment (nihaals) [inline packages/components/package.json:5]: From the [docs](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#repository):
> The URL should be a publicly available (perhaps read-only) url that can be handed directly to a VCS program without any modification. It should not be a url to an html project page that you put in your browser. It's for computers.

A relevant example:
```json
{
  "repository": {
    "type": "git",
    "url": "https://github.com/facebook/react.git",
    "directory": "packages/react-dom"
  }
}
```

File: packages/storybook/package.json [hunk 8]
```diff
@@ -0,0 +1,24 @@
+{
+  "name": "@thunderstore/storybook",
+  "version": "0.1.0",
+  "private": true,
+  "repository": "https://github.com/thunderstore-io/thunderstore-ui/tree/master/packages/storybook",
```
Comment (nihaals) [inline packages/storybook/package.json:5]: From the [docs](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#repository):
> The URL should be a publicly available (perhaps read-only) url that can be handed directly to a VCS program without any modification. It should not be a url to an html project page that you put in your browser. It's for computers.

A relevant example:
```json
{
  "repository": {
    "type": "git",
    "url": "https://github.com/facebook/react.git",
    "directory": "packages/react-dom"
  }
}
```

File: package.json [hunk 9]
```diff
@@ -2,25 +2,17 @@
...
-    "thunderstore-components"
+    "packages/*"
   ],
-  "dependencies": {
-    "react": "^17.0.2",
-    "react-dom": "^17.0.2"
-  },
   "devDependencies": {
     "@microsoft/eslint-formatter-sarif": "^2.1.5",
-    "@types/node": "^14.14.37",
```
Comment (nihaals) [inline package.json:14]: This package isn't in any of the `package.json`s anymore

File: packages/components/src/components/PackageUpload.tsx [hunk 10]
```diff
@@ -139,9 +139,15 @@ const PackageUploadForm: React.FC<PackageUploadFormProps> = ({
                       defaultValue={null}
                       control={control}
                     />
-                    {errors.communities?.message && (
-                      <FormErrorMessage>{errors.communities.message}</FormErrorMessage>
-                    )}
+                    {errors.communities &&
+                      errors.communities.map(
```
Comment (nihaals) [inline packages/components/src/components/PackageUpload.tsx:143]: This gives me `TypeError: errors.communities.map is not a function` if I press `Upload` without entering any values (as it is a string, not a list of strings)

File: packages/components/src/constants.ts [hunk 11]
```diff
@@ -0,0 +1,2 @@
+export const isDevelopment = process.env["NODE_ENV"] === "development";
+export const isProduction = process.env["NODE_ENV"] === "production";
```
Comment (nihaals) [inline packages/components/src/constants.ts:2]: What's wrong with `process.env.NODE_ENV`?

File: packages/nextjs/pages/_app.tsx [hunk 12]
```diff
@@ -1,12 +1,12 @@
...
-import { RootWrapper, theme } from "thunderstore-components";
+import { RootWrapper, theme } from "@thunderstore/components";
 
 function MyApp({ Component, pageProps }: AppProps): JSX.Element {
   return (
     <RootWrapper
       theme={theme}
       thunderstoreProviderValue={{
-        apiUrl: process.env.NEXT_PUBLIC_API_URL || "https://thunderstore.io/api/",
+        apiUrl: process.env["NEXT_PUBLIC_API_URL"] || "https://thunderstore.io/api/",
```
Comment (nihaals) [inline packages/nextjs/pages/_app.tsx:9]: What's wrong with `process.env.NEXT_PUBLIC_API_URL`?

**Review by nihaals — CHANGES_REQUESTED** [review]
The `dependabot.yml` file might be outdated now, I'm not sure if it checks yarn workspaces and updates those correctly, especially with some packages being in multiple `package.json`s

GitHub says that `ssri=6.0.1` is used which has a [vulnerability](https://github.com/advisories/GHSA-vx3p-948g-6vhq). There are also mentions of `8.0.1` so I'm not sure which version is actually being used, but should definitely make sure that it's not possible to install `<8.0.1`.

I haven't looked at the `tsconfig`s yet but some of the references to `types/` might be redundant.

`typings/` is also the standard name for a directory like that.

Overall great changes and sets good foundations for the final storybook setup (e.g. msw).

File: package.json [hunk 13]
```diff
@@ -2,25 +2,17 @@
   "private": true,
   "repository": "https://github.com/thunderstore-io/thunderstore-ui",
   "workspaces": [
-    "nextjs",
```
[the remark this answers is not in the document]
Comment (MythicManiac) [reply package.json:5]: I don't think it really matters, keeps the configuration simpler to have everything under the same rules of structuring. If anything you could rename the `packages` to something like `projects`.

File: package.json (hunk 9 above)
Comment (MythicManiac) [reply package.json:14]: all of the typechecks work just fine without it at least

File: packages/components/src/components/PackageUpload.tsx (hunk 10 above)
Comment (MythicManiac) [reply packages/components/src/components/PackageUpload.tsx:143]: In that case the type definition for `errors.communities` was wrong and should be looked at. I just made it pass typechecks here.

File: packages/components/src/constants.ts (hunk 11 above)
Comment (MythicManiac) [reply packages/components/src/constants.ts:2]: lint rule requiring indexes to be accessed as indexes

File: packages/nextjs/pages/_app.tsx (hunk 12 above)
Comment (MythicManiac) [reply packages/nextjs/pages/_app.tsx:9]: same as above

File: packages/storybook/package.json (hunk 8 above)
Comment (MythicManiac) [reply packages/storybook/package.json:5]: Iherited the pattern from the existing two projects, but I'll change this and those to just point to the root repo here

File: packages/components/src/components/PackageUpload.tsx (hunk 10 above)
Comment (nihaals) [reply packages/components/src/components/PackageUpload.tsx:143]: Yeah that's why I left it at the time. I think we might need to make an issue on react-hook-form. It's either a list of strings or a string

File: packages/components/src/constants.ts (hunk 11 above)
Comment (nihaals) [reply packages/components/src/constants.ts:2]: Which lint rule? I don't think I saw an ESLint error

Comment (MythicManiac) [reply packages/components/src/constants.ts:2]: some typescript strict flag

File: package.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (nihaals) [reply package.json:5]: I think `projects` sounds better since we're unlikely to have more libraries and the components are mostly for our own projects. I don't have a strong preference between the two though, it's pretty minor

File: tsconfig.build.json [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+    "noImplicitAny": true,
+    "strictNullChecks": true,
+    "strictFunctionTypes": true,
+    "strictBindCallApply": true,
+    "strictPropertyInitialization": true,
+    "noImplicitThis": true,
+    "alwaysStrict": true,
+
+    "noUnusedLocals": true,
+    "noUnusedParameters": false,
```
Comment (nihaals) [inline tsconfig.build.json:24]: `noUnusedParameters` should be enabled since `ts(6133)` and `@typescript-eslint/no-unused-vars` enforce no unused parameters (and there isn't really a reason to allow it anyway)

File: packages/components/src/components/PackageUpload.tsx (hunk 10 above)
Comment (nihaals) [reply packages/components/src/components/PackageUpload.tsx:143]: `PackageUploadFormInputs.communities` should be changed to `NestedValue<string[]>` as "non-primitive types" should be wrapped with `NestedValue` (the docs aren't very clear). This fixes the type for `errors`

File: package.json (hunk 9 above)
Comment (nihaals) [reply package.json:14]: I think we can just add this if it's needed in the future for `getServerSideProps`

File: tsconfig.build.json (hunk 14 above)
Comment (MythicManiac) [reply tsconfig.build.json:24]: sometimes you might have a callback where you for example need the 3rd parameter, and IMO it's nicer to just name them out instead of having `_, _, something` for example, which is why I figured this could be disabled.

**Comment by anttimaki** [comment]
Replaced by PR #289 .
